### PR TITLE
2.3.x.twkb

### PR DESCRIPTION
**Pull request by pramsey**
## Standard Parameters
- `table`: Either the name of a table to map from or a SQL query to attempt to map (required)
- `geometry_table`: In the case of a SQL query in the `table` parameter, the table in the query that is the source of the geometry
- `geometry_field`: The name of the column in the table or SQL query that will return the geometry
- `key_field`: The name of the column in the table or SQL query that contains a unique key
- `autodetect_key_field`: Try and figure out the keyfield from metadata tables. (default: false)
- `cursor_size`: If non-zero, the maximum number of rows to fetch while processing a query. For memory constrained situations, this limits the amount of data held in memory during query processing. (default: 0)
- `row_limit`: If non-zero, apply a this `LIMIT` to the query, effectively placing a maximum on the number of features to be mapped. (default: 0)
- `srid`: If non-zero, the EPSG SRID to use for the geometries coming from this query, notwithstanding the SRID that is actually on them. (default: 0)
- `host`, `port`, `dbname`, `user`, `password`: PostgreSQL database connection parameters (required)
- `connect_timeout`: Database connection timeout. (default: 4)
- `pool_max_size`: How big to make the connection pool. (default: 10)
- `persist_connection`: Keep connections open between requests? (default: true)
- `estimate_extent`: Try and estimate extent of data? (default: false)
- `extent_from_subquery`: If trying to estimate extent, and SQL is subquery, estimate on that? (default: false)
- `max_async_connection`: If > 1, turn on the asyncronous connection feature. (default: 1)
- `simplify_geometries`: Use PostGIS ST_Simplify call on the geometry to make input data smaller. (default: false)
- `simplify_dp_ratio`: Use a PostGIS simplification factor of this proportion of the ground size of a pixel. For example, if pixels are 40m square for this rendering, and the ratio is set to 1/20, then `ST_Simplify(geom, 2)` would be used. (default: 1/20)
## 2.3.x.twkb Parameters
- `simplify_snap_ratio`: Use PostGIS `ST_SnapToGrid` call to make the input data smaller. This is applied to geometry **before** any `ST_Simplify` is called, so should use a tolerance smaller than the simplify tolerance. Tolerance is expressed as a proportion of a pixel, as with `simplify_dp_ratio`. (default: 1/40)

Features below all require PostGIS 2.2+, as they use features that do not appear in earlier PostGIS versions.
- `twkb_encoding`: Use TWKB to encode geometries for transport from the database to the renderer instead of standard WKB? (default: false)
- `simplify_dp_preserve`: Set the `preserve` option in `ST_Simplify` when in `simplify_geometries` mode? This will ensure that features that get simplified down to nothing aren't dropped but are retained in point-like form. Useful for rendering to avoid gaps where small feature "disappear". (default: false)
- `simplify_clip_resolution`: In non-zero, sets the map scale at which geometries start getting clipped to the rendering window. (default: 0.0)
## SQL Tokens

SQL tokens are strings that are replaced in the input SQL before the SQL is sent to the database for execution. This is information that Mapnik knows as the result of the request, and can be useful in forming complex SQL queries.
- `!bbox!`
- `!scale_denominator!`
- `!pixel_width!`
- `!pixel_height!`

### CODE REVIEW

**Comment by pramsey** [comment]
per request from @javisantana, a PR for @rochoa and @zenitraM to consider merging into our base 2.3.x branch with doco. For better doco management, I put a [README.md](https://github.com/CartoDB/mapnik/blob/a94ac4e160ba806741d38e043fa26f3f4261789b/plugins/input/postgis/README.md) right into the driver directory too.

**Comment by pramsey** [comment]
Oh, hey, in terms of maintaining our own branches in our repo, do we want to have cartodb/mapnik/2.3.x be our own 2.3.x or do we want it to mirror the mapnik/2.3.x? If we're going to run a patched mapnik, would it makes sense to have a 2.3.x.cartodb branch to which all our customizations go?

**Comment by javisantana** [comment]
@rochoa do we want to continue with 2.3.x in our repo or do we move to 2.3.x.cartodb so it's explicit we are working with a different version from the mapbox one

**Comment by rochoa** [comment]
I would try to get all our changes into mapnik/`2.3.x` so everyone else can benefit from it, however we need to track the differences. So I guess one option is to use `2.3.x-cdb` or something similar which will be the one we use until changes are accepted into mapnik repo. If they don't get there then we will need to keep with the patched branch.

What do you think?

**Comment by rochoa** [comment]
I'm gonna update our 2.3.x branch with upstream so this PR will be simpler.

File: plugins/input/postgis/postgis_datasource.cpp [hunk 1]
```diff
@@ -482,21 +507,18 @@ std::string postgis_datasource::sql_bbox(box2d<double> const& env) const
...
-    if (srid_ > 0)
-    {
-        b << "ST_SetSRID(";
-    }
-
-    b << "'BOX3D(";
+    b << "ST_MakeEnvelope(";
     b << std::setprecision(16);
-    b << env.minx() << " " << env.miny() << ",";
-    b << env.maxx() << " " << env.maxy() << ")'::box3d";
```
Comment (rochoa) [inline plugins/input/postgis/postgis_datasource.cpp:493]: No need to use `::box3d`?

**Comment by rochoa** [comment]
How hard would be to add some simple tests? :smiling_imp:

**Comment by pramsey** [comment]
I'm closing this in favor of #2, which targets our private features branch, 2.3.x.cartodb.

File: plugins/input/postgis/postgis_datasource.cpp (hunk 1 above)
Comment (pramsey) [inline plugins/input/postgis/postgis_datasource.cpp:493]: No, going through box3d is actually a little more involved (build box3d type, cast to geometry, then use) than building a geometry in one go with `ST_MakeEnvelope()`

**Comment by pramsey** [comment]
Regarding tests, I could potentially add some tests to [postgis_test.py](https://github.com/CartoDB/mapnik/blob/1b0f4d663cb0ac51f6c2fef351de888834f30367/tests/python_tests/postgis_test.py) I guess, similar to [this](https://github.com/CartoDB/mapnik/blob/1b0f4d663cb0ac51f6c2fef351de888834f30367/tests/python_tests/postgis_test.py#L789)

**Pull request closed without merging**